shelley/chain-and-ledger/executable-spec/src/PoolRank.hs

```diff
@@ -0,0 +1,93 @@
+module PoolRank
+  ( desirability
+  , rankedPools
+  ) where
+
+import           BaseTypes (intervalValue)
+import           Coin (Coin (..))
+import           Data.Map.Strict (Map)
+import qualified Data.Map.Strict as Map
+import           Data.Ratio ((%))
+import qualified Data.Set as Set
+import           EpochBoundary (BlocksMade (..), SnapShots (..), Stake (..), poolStake)
+import           Keys (KeyHash)
+import           LedgerState (AccountState (..), EpochState (..), NewEpochState (..))
+import           Numeric.Natural (Natural)
+import           PParams (PParams (..))
+import           Slot (EpochSize)
+import           TxData (PoolParams (..))
+
+
+-- | Desirability calculation for non-myopic utily,
+-- corresponding to f^~ in section 5.6.1 of
+-- "Design Specification for Delegation and Incentives in Cardano"
+desirability
+  :: PParams
+  -> Coin
+  -> Natural
+  -> Natural
+  -> PoolParams crypto
+  -> Stake crypto
+  -> Coin
+  -> Rational
+desirability pp r blocksN blocksTotal pool (Stake stake) (Coin total) =
+  if fTilde <= cost
+    then 0
+    else (fTilde - cost)*(1 -margin)
+  where
+    fTilde      = fTildeNumer / fTildeDenom
+    fTildeNumer = pBar * fromIntegral r * (z0 + min s z0 * a0)
+    fTildeDenom = 1 + a0
+
+    cost   = fromIntegral $ _poolCost pool
+    margin = intervalValue $ _poolMargin pool
+    tot = max 1 (fromIntegral total)
+
+    Coin pledge = _poolPledge pool
+    s = fromIntegral pledge % tot
+    a0 = _a0 pp
+    z0 = 1 % max 1 (fromIntegral (_nOpt pp))
+
+    pBar = if intervalValue (_d pp) < 0.8 && sigma > 0
```

Comment:
How do we want to handle `sigma == 0` pools? I would argue that it would be better to set their apparent performance to zero than to one, since `sigma == 0` implies that the pool doesn't even have a pledge.

Reply:
that's a good point, I agree, zero makes more sense.

I was thinking that it didn't matter what the apparent performance was if there was not stake, but I think this reasoning only makes sense for rewards, not ranking.

If a pool has no stake, then they would never be able to produce a block, so performance is does not matter for the reward calculation (either the pledge is zero, so `f(s, sigma)` is zero, or the pledge has not been met).

But for ranking, I agree, 0 makes more sense. And I like the idea of being consistent, so how about we do the same thing for both rewards and ranking.

Reply:
I went with 1 (when I was thinking that it didn't matter), since "great!" seemed like the correct answer to the question "how well did you do at producing zero blocks?". :grin: